File: 2434-design-a-number-container-system/2434-design-a-number-container-system.cpp
class NumberContainers {
public:
    NumberContainers() {}

    void change(int index, int number) {
        indexToNumbers[index] = number;

        numberToIndices[number].push(index);
    }

    int find(int number) {
        if (numberToIndices.find(number) == numberToIndices.end()) {
            return -1;
        }

        auto& minHeap = numberToIndices[number];

        while (!minHeap.empty()) {
            int index = minHeap.top();

            if (indexToNumbers[index] == number) {
                return index;
            }

            minHeap.pop();
        }

        return -1;
    }

private:
    unordered_map<int, priority_queue<int, vector<int>, greater<int>>>
        numberToIndices;
    unordered_map<int, int> indexToNumbers;
};

File: 2434-design-a-number-container-system/2434-design-a-number-container-system.java
class NumberContainers {

    public NumberContainers() {
        numberToIndices = new HashMap<>();
        indexToNumbers = new HashMap<>();
    }

    public void change(int index, int number) {
        indexToNumbers.put(index, number);

        numberToIndices
            .computeIfAbsent(number, k -> new PriorityQueue<>())
            .add(index);
    }

    public int find(int number) {
        if (!numberToIndices.containsKey(number)) {
            return -1;
        }

        PriorityQueue<Integer> minHeap = numberToIndices.get(number);

        while (!minHeap.isEmpty()) {
            int index = minHeap.peek();

            if (indexToNumbers.get(index) == number) {
                return index;
            }

            minHeap.poll();
        }

        return -1;
    }

    private Map<Integer, PriorityQueue<Integer>> numberToIndices;
    private Map<Integer, Integer> indexToNumbers;
}